exporter/influxdbexporter/writer.go

```diff
@@ -0,0 +1,179 @@
+	"go.opentelemetry.io/collector/consumer/consumererror"
+)
+
+type influxHTTPWriter struct {
+	encoderPool sync.Pool
+	httpClient  *http.Client
+	writeURL    string
+	writeHeader http.Header
+
+	logger common.Logger
+}
+
+func newInfluxHTTPWriter(logger common.Logger, serverURL, authToken, org, bucket string, writeTimeout time.Duration) (*influxHTTPWriter, error) {
+	writeURL, err := url.Parse(serverURL)
+	if err != nil {
+		return nil, err
+	}
+	writeURL, err = writeURL.Parse("api/v2/write")
+	if err != nil {
+		return nil, err
+	}
+	queryValues := writeURL.Query()
+	queryValues.Set("org", org)
+	queryValues.Set("bucket", bucket)
+	queryValues.Set("precision", "ns")
+	writeURL.RawQuery = queryValues.Encode()
+
+	writeHeader := make(http.Header)
+	writeHeader.Set("User-Agent", "OpenTelemetry -> Influx")
+	if authToken != "" {
+		writeHeader.Set("Authorization", "Token "+authToken)
+	}
+
+	httpClient := new(http.Client)
+	if writeTimeout > 0 {
+		httpClient.Timeout = writeTimeout
+	}
+
+	return &influxHTTPWriter{
+		encoderPool: sync.Pool{
+			New: func() interface{} {
+				e := new(lineprotocol.Encoder)
+				e.SetLax(false)
+				e.SetPrecision(lineprotocol.Nanosecond)
+				return e
+			},
+		},
+		httpClient:  httpClient,
+		writeURL:    writeURL.String(),
+		writeHeader: writeHeader,
+		logger:      logger,
+	}, nil
+}
+
+func (w *influxHTTPWriter) newBatch() *influxHTTPWriterBatch {
+	return &influxHTTPWriterBatch{
+		w:       w,
+		encoder: w.encoderPool.Get().(*lineprotocol.Encoder),
+		logger:  w.logger,
+	}
+}
+
+type influxHTTPWriterBatch struct {
+	w       *influxHTTPWriter
+	encoder *lineprotocol.Encoder
+	logger  common.Logger
+}
+
+func (b *influxHTTPWriterBatch) WritePoint(_ context.Context, measurement string, tags map[string]string, fields map[string]interface{}, ts time.Time) error {
+	b.encoder.StartLine(measurement)
+	for _, tag := range b.sortTags(tags) {
+		b.encoder.AddTag(tag.k, tag.v)
+	}
+	for k, v := range b.convertFields(fields) {
+		b.encoder.AddField(k, v)
+	}
+	b.encoder.EndLine(ts)
+
+	if err := b.encoder.Err(); err != nil {
+		// b.encoder.AbortLine
+		defer b.encoder.ClearErr()
+		return fmt.Errorf("failed to encode point: %w", err)
+	}
+
+	return nil
+}
+
+func (b *influxHTTPWriterBatch) flushAndClose(ctx context.Context) error {
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, b.w.writeURL, bytes.NewReader(b.encoder.Bytes()))
+	if err != nil {
+		return consumererror.Permanent(err)
+	}
+	req.Header = b.w.writeHeader
+
+	if res, err := b.w.httpClient.Do(req); err != nil {
+		return err
+	} else if err = res.Body.Close(); err != nil {
+		return err
+	} else {
+		switch res.StatusCode / 100 {
+		case 2: // Success
+			break
+		case 5: // Retryable error
+			return fmt.Errorf("line protocol write returned %q", res.Status)
+		default: // Terminal error
+			return consumererror.Permanent(fmt.Errorf("line protocol write returned %q", res.Status))
```

Comment:
If there is a body in the response that explains the problem consider including it. It may help with troubleshooting since this error is printed in the logs by queued_retry.